File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeList(ListNode* list1, ListNode* list2){
        if(list1==nullptr)  return list2;
        if(list2==nullptr)  return list1;
        if(list1->val<list2->val){
            list1->next = mergeList(list1->next, list2);
            return list1;
        }else{
            list2->next = mergeList(list1, list2->next);
            return list2;
        }
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return nullptr;
        if(lists.size()==1) return lists[0];
        ListNode* last=lists.back();    lists.pop_back();
        ListNode* last2=lists.back();   lists.pop_back();
        ListNode* temp = mergeList(last, last2);
        lists.push_back(temp);
        return mergeKLists(lists);
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeList(self, list1, list2):
        if not list1:
            return list2
        if not list2:
            return list1
        if list1.val < list2.val:
            list1.next = self.mergeList(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeList(list1, list2.next)
            return list2
        
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return
        if len(lists) == 1:
            return lists[0]
        
        last  = lists.pop()
        last2 = lists.pop()
        
        lists.append( self.mergeList(last, last2) )
        
        return self.mergeKLists(lists)
        